### PR TITLE
Make branch build again.

### PR DESCRIPTION
**Pull request by jeschkies**
* Don't write empty collections to Json.
* Deal with empty `acceptedResourceRoles` collection in proto write.
* Check for non empty tasks in `TaskKiller`.
* Don't treat storage as future but as option.
* Correct `RunSpec` reference in docstrings.
* Map task killer answer correctly.

### CODE REVIEW

File: src/main/scala/mesosphere/marathon/storage/migration/Migration.scala [hunk 1]
```diff
@@ -173,16 +173,17 @@ class Migration(
...
+        legacyStore.map { store =>
+          store.load(StorageVersionName).flatMap {
+            case Some(entity) => store.update(entity.withNewContent(bytes))
+            case None => store.create(StorageVersionName, bytes)
+          }
         }
-      }
-    }.getOrElse(Future.successful(Done))
-    await(storeVersionFuture)
+    }
```
Comment (jeschkies) [inline src/main/scala/mesosphere/marathon/storage/migration/Migration.scala]: `persistenceStore` is not a future as suggested originally. This change should make it more explicit. Do we want a future instead?

Comment (jdef) [inline src/main/scala/mesosphere/marathon/storage/migration/Migration.scala]: /cc @jasongilanfarr

Comment (jasongilanfarr) [inline src/main/scala/mesosphere/marathon/storage/migration/Migration.scala]: setting the storage version is a future and we should be waiting on it.

File: src/main/scala/mesosphere/marathon/api/v2/json/Formats.scala [hunk 2]
```diff
@@ -1010,13 +1009,21 @@ trait AppAndGroupFormats {
...
         "labels" -> runSpec.labels,
-        "acceptedResourceRoles" -> runSpec.acceptedResourceRoles,
         "ipAddress" -> runSpec.ipAddress,
         "version" -> runSpec.version,
         "residency" -> runSpec.residency,
         "secrets" -> runSpec.secrets,
         "taskKillGracePeriodSeconds" -> runSpec.taskKillGracePeriod
       )
+
+      if (runSpec.args.nonEmpty) {
```
Comment (jasongilanfarr) [inline src/main/scala/mesosphere/marathon/api/v2/json/Formats.scala]: Don't see why this matters

File: src/main/scala/mesosphere/marathon/state/AppDefinition.scala [hunk 3]
```diff
@@ -163,9 +163,11 @@ case class AppDefinition(
     readinessChecks.foreach { r => builder.addReadinessCheckDefinition(ReadinessCheckSerializer.toProto(r)) }
     taskKillGracePeriod.foreach { t => builder.setTaskKillGracePeriod(t.toMillis) }
 
-    val roles = Protos.ResourceRoles.newBuilder()
-    roles.addAllRole(acceptedResourceRoles.asJava)
-    builder.setAcceptedResourceRoles(roles)
+    if (acceptedResourceRoles.nonEmpty) {
+      val roles = Protos.ResourceRoles.newBuilder()
```
Comment (jasongilanfarr) [inline src/main/scala/mesosphere/marathon/state/AppDefinition.scala]: Why would this matter?

Comment (jeschkies) [inline src/main/scala/mesosphere/marathon/state/AppDefinition.scala]: If we set an empty list proto will set an empty list and `hasAcceptedResourceRoles` will be true. It broke this assertion: https://github.com/mesosphere/marathon/blob/feature/pods/src/test/scala/mesosphere/marathon/state/AppDefinitionTest.scala#L72.

Comment (jasongilanfarr) [inline src/main/scala/mesosphere/marathon/state/AppDefinition.scala]: We should fix that then... hasAcceptedResourceRoles should be == "acceptedResourceRoles.nonEmpty"

File: src/main/scala/mesosphere/marathon/storage/migration/Migration.scala (hunk 1 above)
Comment (jeschkies) [inline src/main/scala/mesosphere/marathon/storage/migration/Migration.scala]: Okay. Changed it.

File: src/main/scala/mesosphere/marathon/state/AppDefinition.scala (hunk 3 above)
Comment (jasongilanfarr) [inline src/main/scala/mesosphere/marathon/state/AppDefinition.scala]: If we don't set an empty list on the protos, when we read it, it'll still be empty ;)

File: src/main/scala/mesosphere/marathon/api/v2/json/Formats.scala (hunk 2 above)
Comment (jeschkies) [inline src/main/scala/mesosphere/marathon/api/v2/json/Formats.scala]: It'll be an empty list and break this assetion 
`(r1 \ "args").asOpt[Seq[String]] should equal (None)` in
https://github.com/mesosphere/marathon/blob/feature/pods/src/test/scala/mesosphere/marathon/api/v2/json/AppDefinitionFormatsTest.scala#L53

File: src/main/scala/mesosphere/marathon/state/AppDefinition.scala (hunk 3 above)
Comment (jeschkies) [inline src/main/scala/mesosphere/marathon/state/AppDefinition.scala]: Hehe, yeah. However, `hasAcceptedResourceRoles` looks for `null`. I don't have a strong opinion on this. I just didn't want to change the assertions.

File: src/main/scala/mesosphere/marathon/api/v2/json/Formats.scala (hunk 2 above)
Comment (jasongilanfarr) [inline src/main/scala/mesosphere/marathon/api/v2/json/Formats.scala]: Reasonable, I guess- I still think it's awkward to treat empty list and unset lists differently.

Comment (jeschkies) [inline src/main/scala/mesosphere/marathon/api/v2/json/Formats.scala]: Me too. It should default to an empty list. Let me fix that then.

Comment (jeschkies) [inline src/main/scala/mesosphere/marathon/api/v2/json/Formats.scala]: Hm, so it seems we do have a semantic difference between `Set.empty` and `null` for accepted resource roles: https://github.com/mesosphere/marathon/blob/master/src/test/scala/mesosphere/marathon/api/v2/json/AppDefinitionFormatsTest.scala#L187

At least we expect an exception.

**Comment by jasongilanfarr** [comment]
(merged)

**Pull request closed without merging**